File: shuffle-service/src/main/scala/org/apache/spark/shuffle/ytsaurus/ShuffleUtils.scala
package org.apache.spark.shuffle.ytsaurus

import org.apache.spark.SparkConf
import org.apache.spark.shuffle.ytsaurus.Config.YTSAURUS_SHUFFLE_CONFIG
import tech.ytsaurus.ysontree.{YTreeMapNode, YTreeTextSerializer}

import scala.jdk.CollectionConverters._

object ShuffleUtils {

  private val PULL_SECTION = "pull"
  private val PUSH_SECTION = "push"

  def shuffleConfig(conf: SparkConf, pushBasedEnabled: Boolean): Option[YTreeMapNode] = {
    val config = conf.get(YTSAURUS_SHUFFLE_CONFIG).map(YTreeTextSerializer.deserialize(_).mapNode())
    config.foreach(validateModeSection(_, pushBasedEnabled))
    config
  }

  private def validateModeSection(config: YTreeMapNode, pushBasedEnabled: Boolean): Unit = {
    val activeSection = if (pushBasedEnabled) PUSH_SECTION else PULL_SECTION
    val unexpectedKeys = config.keys().asScala.filterNot(_ == activeSection)
    require(unexpectedKeys.isEmpty,
      s"${YTSAURUS_SHUFFLE_CONFIG.key} may contain only the '$activeSection' section when push-based shuffle is " +
        s"${if (pushBasedEnabled) "enabled" else "disabled"}, got unexpected keys: ${unexpectedKeys.mkString(", ")}")
  }
}

File: shuffle-service/src/main/java/org/apache/spark/shuffle/ytsaurus/ShuffleUtils.java
package org.apache.spark.shuffle.ytsaurus;

import org.apache.spark.SparkConf;

import tech.ytsaurus.ysontree.YTreeMapNode;
import tech.ytsaurus.ysontree.YTreeTextSerializer;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import scala.Option;

import static org.apache.spark.shuffle.ytsaurus.Config.YTSAURUS_SHUFFLE_CONFIG;

public final class ShuffleUtils {
    private ShuffleUtils() { }

    static <T> Optional<T> toJavaOptional(Option<T> value) {
        return value.isDefined() ? Optional.ofNullable(value.get()) : Optional.empty();
    }

    public static Optional<YTreeMapNode> shuffleConfig(SparkConf conf, boolean pushBasedEnabled) {
        Optional<String> value = toJavaOptional(conf.get(YTSAURUS_SHUFFLE_CONFIG));
        if (value.isEmpty()) {
            return Optional.empty();
        }
        YTreeMapNode config = YTreeTextSerializer.deserialize(value.get()).mapNode();
        String activeSection = pushBasedEnabled ? "push" : "pull";
        List<String> unexpectedKeys = config.keys().stream()
                .filter(key -> !key.equals(activeSection))
                .collect(Collectors.toList());
        if (!unexpectedKeys.isEmpty()) {
            throw new IllegalArgumentException("requirement failed: " + YTSAURUS_SHUFFLE_CONFIG.key()
                    + " may contain only the '" + activeSection + "' section when push-based shuffle is "
                    + (pushBasedEnabled ? "enabled" : "disabled") + ", got unexpected keys: "
                    + String.join(", ", unexpectedKeys));
        }
        return Optional.of(config);
    }
}
